src/lib/pubkey/dilithium/dilithium_common/dilithium_algos.cpp

```diff
@@ -516,12 +526,15 @@ DilithiumPoly sample_in_ball(StrongSpan<const DilithiumCommitmentHash> seed, con
    auto xof = mode.symmetric_primitives().H(seed);
 
    // This generator resembles the while loop in the spec.
-   auto next_byte_lower_than = [&xof](size_t i) -> uint8_t {
-      while(true) {
+   auto next_byte_lower_than = [&xof, n = uint16_t(0)](size_t i) mutable -> uint8_t {
+      while(n < DilithiumConstants::SAMPLE_IN_BALL_XOF_BOUND) {
+         ++n;
          if(const uint8_t b = xof.output_next_byte(); b <= i) {
             return b;
          }
       }
+
+      throw Internal_Error("SampleInBall did not terminate");
```

Comment:
Is there optimization potential since we know we only need 221 XOF bytes? We could create an array of size SAMPLE_IN_BALL_XOF_BOUND and fill it with the XOF in one go. Do you think that would be faster?